### PR TITLE
Improve DETR post-processing methods

### PR DESCRIPTION
**Pull request by alaradirik**
# What does this PR do?
- Adds `post_process_object_detection`, `post_process_semantic_segmentation`, `post_process_instance_segmentation` and `post_process_panoptic_segmentation` methods with optional resizing and thresholding to filter out low probability predictions.

This PR is part of a larger effort to ensure post-processing methods have consistent naming, input arguments and output. Deprecation warnings are added to existing post-processing methods.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/models/detr/feature_extraction_detr.py [hunk 1]
```diff
@@ -939,3 +1005,352 @@ def get_ids_area(masks, scores, dedup=False):
...
+        self, outputs, object_detection_threshold: float = 0.5, target_sizes: Union[TensorType, List[Tuple]] = None
+    ):
+        """
+        Converts the output of [`DetrForObjectDetection`] into the format expected by the COCO api. Only supports
+        PyTorch.
+
+        Args:
+            outputs ([`DetrObjectDetectionOutput`]):
+                Raw outputs of the model.
+            object_detection_threshold (`float`, *optional*):
```
Comment (sgugger) [inline src/transformers/models/detr/feature_extraction_detr.py:1019]: As mentioned by Niels on the internal doc, just `threshold` is enough here. The method is named `post_process_object_detection` so it's obvious we are doing this.

File: src/transformers/models/detr/feature_extraction_detr.py [hunk 2]
```diff
@@ -939,3 +1005,352 @@ def get_ids_area(masks, scores, dedup=False):
...
+        else:
+            semantic_segmentation = segmentation.argmax(dim=1)
+            semantic_segmentation = [semantic_segmentation[i] for i in range(semantic_segmentation.shape[0])]
+
+        return semantic_segmentation
+
+    def post_process_instance_segmentation(
+        self,
+        outputs,
+        object_mask_threshold: float = 0.5,
```
Comment (sgugger) [inline src/transformers/models/detr/feature_extraction_detr.py:1112]: Same here.

**Review by sgugger — APPROVED** [review]
Thanks for working on this. As mentioned internally, don't hesitate to separate PRs in focused bits: here there could have been:
- one PR to quickly fix the comments
- one PR to add the post processing to DETR
- one PR for the deprecation warnings

Also wondering if we should target a lower version than v5 for the removal of the methods, since this is all a bit of an experimental API.

**Comment by alaradirik** [comment]
@amyeroberts @sgugger thank you both for the feedback! Could you take a final look at the PR? If everything looks fine, I'll apply the same changes to MaskFormer and update the model cards.

**Comment by alaradirik** [comment]
> Thanks for working on this. As mentioned internally, don't hesitate to separate PRs in focused bits: here there could have been:
> 
> * one PR to quickly fix the comments
> * one PR to add the post processing to DETR
> * one PR for the deprecation warnings

Makes sense, I'll split my future PRs into smaller, issue-specific ones.

> Also wondering if we should target a lower version than v5 for the removal of the methods, since this is all a bit of an experimental API.

I agree, we can target a lower version for the removal of deprecated methods.

File: src/transformers/models/detr/feature_extraction_detr.py [hunk 3]
```diff
@@ -400,6 +400,52 @@ def _normalize(self, image, mean, std, target=None):
 
         return image, target
 
+    def binary_mask_to_rle(self, mask):
```
Comment (amyeroberts) [inline src/transformers/models/detr/feature_extraction_detr.py:403]: I think both of these methods `binary_mask_to_rle` and `remove_low_and_no_objects` should be moved out of the class to be top-level functions: they are helper functionalities used by the feature extractor to process.

**Review by amyeroberts — APPROVED** [review]
LGTM 👍  Thanks for the PR and work on improving this

Left a few nits, main comment is about moving the processing helper functions out of the class.

File: src/transformers/models/conditional_detr/feature_extraction_conditional_detr.py [hunk 4]
```diff
@@ -753,6 +753,11 @@ def post_process_segmentation(self, outputs, target_sizes, threshold=0.9, mask_t
             `List[Dict]`: A list of dictionaries, each dictionary containing the scores, labels, and masks for an image
             in the batch as predicted by the model.
         """
+        logger.warning(
+            "`post_process_segmentation` is deprecated and will be removed in v5 of Transformers, please use"
+            " `post_process_semantic_segmentation`.",
```
[the remark this answers is not in the document]
Comment (alaradirik) [reply src/transformers/models/conditional_detr/feature_extraction_conditional_detr.py:758]: @amyeroberts exactly, I will open separate PRs for Conditional DETR and Deformable DETR.

**Pull request merged**